### DISCUSSION TITLE
Can someone give me an exmaple of promoting a piece.

### DISCUSSION THREAD

**Discussion by f1forhelp** [alexobviously/bishop#62] [Q&A] [answered]
Can someone give me an exmaple of promoting a piece.

I am basically inserting pieces one by one but when pawn reaches at the end I want to show the standard promotion options and if selected make the change in game.

``
    Move? move = game.getMove(
        "${initialData.fileLetter}${initialData.rankNumber}${finalData.fileLetter}${finalData.rankNumber}");
        
    if (move == null) {
      return;
    }
    
    game.makeMove(mov);

I initialized using the standard variant .

  late Game game;

  ChessController() {
    game = Game(variant: Variant.standard());
  }

**Comment by alexobviously** [comment]
You put the piece symbol at the end of the algebraic move string. So like `a7a8` would be moving a piece from a7 to a8, but if that's a pawn and you want to promote to a queen it should be `a7a8q`. You can also do `a7a8n`, `a7a8b` or `a7a8r` for knight, bishop or rook, respectively.

By the way, you can list all the moves with `game.generateLegalMoves();`, for debugging if you want. There's also `game.algebraicMoves();` which will give you all the moves in algebraic string form (this is just a convenience shortcut for `game.generateLegalMoves().toAlgebraic();`). There are also filters you can use like:
```dart
List<Move> moves = game.generateLegalMoves();
List<Move> promos = moves.promoMoves; // this will filter to only promotion moves
int square = game.size.squareNumber('a8'); // I know this is a bit verbose and I might make it easier at some point, it's like this because bishop supports different board sizes
List<Move> squareMoves = moves.to(square);
List<Move> promoSquareMoves = moves.to(square).promoMoves; // the filters are also chainable
List<String> stringMoves = promoSquareMoves.toAlgebraic();
print(stringMoves);
```

**Reply by f1forhelp** [reply]
Cant, we do it using    ` game.makeMove(mov); though `  `var mov = getMove(initialCordinate, finalCordinate); is` throwing me a null response. when trying to move the pawn after it reaches at the end.
And from your solution will it be saving its state?

Obviously it should not move after it reaches end but then how can I make a promotion before it moves to end and also save the promotion to game state.

**Reply by alexobviously** [reply, marked as answer]
Well you said you're calling it like this: `Move? move = game.getMove(
"${initialData.fileLetter}${initialData.rankNumber}${finalData.fileLetter}${finalData.rankNumber}");`

What you need to do is:
```dart
String m = '${initialData.fileLetter}${initialData.rankNumber}${finalData.fileLetter}${finalData.rankNumber}';
if(initialData.promoPiece!=null) {
  m = '$m${initialData.promoPiece}';
}
final move = game.getMove(m);
```

**Reply by alexobviously** [reply]
Oh I think I understand, you want to break the promotion move into two steps. That's something you have to do on the UI side, because 'pawn moves to a8 and promotes to queen' is one move, not two.

By the way, if you're doing something in flutter, you can just use [squares](https://pub.dev/packages/squares) which handles this sort of thing for you.

If you want to do it yourself, you can also look at [how squares does it](https://github.com/alexobviously/squares/blob/main/lib/src/ui/board_controller.dart#L301) for a bit of inspiration. But essentially what you want is:
* a board controller that knows all the possible moves
* when a user moves a piece to a square, you check this list of possible moves
* if there are possible promotions on that move, you store the partial move (from and to squares), and show a piece selector
* when the piece selector returns a value, you use that to complete the move and send it to the game controller (bishop)
* if the piece selector is closed, you revert the partial move

**Reply by f1forhelp** [reply]
Tnx man now i got your point.

**Reply by alexobviously** [reply]
no problem mate, let me know if you need anything else
